lib/pac55xx/ccs.c

```diff
@@ -0,0 +1,271 @@
+
+void ccs_frclkmux_select(uint32_t sel) {
+	CCSCTL = (CCSCTL & ~CCS_CTL_FRCLKMUXSEL(CCS_CTL_FRCLKMUXSEL_MASK)) | CCS_CTL_FRCLKMUXSEL(sel);
+}
+void ccs_rosc_enable(void) {
+	CCSCTL |= CCS_CTL_ROSCEN;
+}
+void ccs_rosc_disable(void) {
+	CCSCTL &= ~CCS_CTL_ROSCEN;
+}
+void ccs_sclkmux_select_frclk(void) {
+	CCSCTL &= ~CCS_CTL_SCLKMUXSEL;
+}
+void ccs_sclkmux_select_pllclk(void) {
+	CCSCTL |= CCS_CTL_SCLKMUXSEL;
+}
+void ccs_clkfail_enable(void) {
+	CCSCTL |= CCS_CTL_CLKFAILEN;
+}
+void ccs_clkfail_disable(void) {
+	CCSCTL &= ~CCS_CTL_CLKFAILEN;
+}
+void ccs_clkfailmux_select_frclk(void) {
+	CCSCTL &= ~CCS_CTL_CLKFAILMUXSEL;
+}
+void ccs_clkfailmux_select_pllclk(void) {
+	CCSCTL |= CCS_CTL_CLKFAILMUXSEL;
+}
+void ccs_ldo_enable(void) {
+	CCSCTL |= CCS_CTL_LDOEN;
+}
+void ccs_ldo_disable(void) {
+	CCSCTL &= ~CCS_CTL_LDOEN;
+}
+void ccs_pclk_enable(void) {
+	CCSCTL |= CCS_CTL_PCLKEN;
+}
+void ccs_pclk_disable(void) {
+	CCSCTL &= ~CCS_CTL_PCLKEN;
+}
+void ccs_aclk_enable(void) {
+	CCSCTL |= CCS_CTL_ACLKEN;
+}
+void ccs_aclk_disable(void) {
+	CCSCTL &= ~CCS_CTL_ACLKEN;
+}
+void ccs_adcclk_enable(void) {
+	CCSCTL |= CCS_CTL_ADCCLKEN;
+}
+void ccs_adcclk_disable(void) {
+	CCSCTL &= ~CCS_CTL_ADCCLKEN;
+}
+void ccs_stclk_sleep_enable(void) {
+	CCSCTL |= CCS_CTL_STCLKSLPEN;
+}
+void ccs_stclk_sleep_disable(void) {
+	CCSCTL &= ~CCS_CTL_STCLKSLPEN;
+}
+void ccs_set_pclkdiv(uint32_t div) {
+	CCSCTL = (CCSCTL & ~CCS_CTL_PCLKDIV(CCS_CTL_PCLKDIV_MASK)) | CCS_CTL_PCLKDIV(div);
+}
+void ccs_set_aclkdiv(uint32_t div) {
+	CCSCTL = (CCSCTL & ~CCS_CTL_ACLKDIV(CCS_CTL_ACLKDIV_MASK)) | CCS_CTL_ACLKDIV(div);
+}
+void ccs_set_hclkdiv(uint32_t div) {
+	CCSCTL = (CCSCTL & ~CCS_CTL_HCLKDIV(CCS_CTL_HCLKDIV_MASK)) | CCS_CTL_HCLKDIV(div);
+}
+void ccs_pll_enable(void) {
+	CCSPLLCTL |= CCS_PLLCTL_PLLEN;
+}
+void ccs_pll_disable(void) {
+	CCSPLLCTL &= ~CCS_PLLCTL_PLLEN;
+}
+bool ccs_pll_locked(void) {
+	return (CCSPLLCTL & CCS_PLLCTL_PLLLOCK) == CCS_PLLCTL_PLLLOCK;
+}
+void ccs_pll_bypass_enable(void) {
+	CCSPLLCTL |= CCS_PLLCTL_PLLBP;
+}
+void ccs_pll_bypass_disable(void) {
+	CCSPLLCTL &= ~CCS_PLLCTL_PLLBP;
+}
+void ccs_pll_set_outdiv(uint32_t div) {
+	CCSPLLCTL = (CCSPLLCTL & ~CCS_PLLCTL_PLLOUTDIV(CCS_PLLCTL_PLLOUTDIV_MASK)) | CCS_PLLCTL_PLLOUTDIV(div);
+}
+void ccs_pll_set_indiv(uint32_t div) {
+	if (div <= 15 && div >= 1) {
+		CCSPLLCTL = (CCSPLLCTL & ~CCS_PLLCTL_PLLINDIV(CCS_PLLCTL_PLLINDIV_MASK)) | CCS_PLLCTL_PLLINDIV(div);
+	}
+}
+void ccs_pll_set_fbdiv(uint32_t div) {
+	if (div <= 16383 && div >= 4) {
+		CCSPLLCTL = (CCSPLLCTL & ~CCS_PLLCTL_PLLFBDIV(CCS_PLLCTL_PLLFBDIV_MASK)) | CCS_PLLCTL_PLLFBDIV(div);
+	}
+}
+void css_pll_config_enable(uint32_t indiv, uint32_t fbdiv, uint32_t outdiv) {
+	ccs_pll_disable();
+	ccs_pll_set_fbdiv(fbdiv);
+	ccs_pll_set_outdiv(outdiv);
+	ccs_pll_set_indiv(indiv);
+	ccs_pll_enable();
+	while (!ccs_pll_locked()) ; /* Wait for PLL lock ~500us */
+}
+uint32_t ccs_get_peripheral_clk_freq(uint32_t periph, uint32_t select) {
```

Comment:
A couple of comments...

1. Depending on outcome of #1191 we may want to split this up to be consistent. 
1. I would suggest that the code here determine the `select` value for the peripheral and not rely on it being passed in so that the user doesn't need to know the details of how the clocks are/have been configured.

Reply:
For now, resolving as okay as is, and both can be fixed pretty easily later, especially since limited use of this at this point.

Reply:
The `select` is used since several of the peripherals choose their own clock within their register set.  If we want this to operate as described in (2) above, then ccs.c will have dependencies on each peripheral's register definition, or at least a dumbed-down version of it.  Not saying it's impossible, just wouldn't look very pretty to have each of those register definitions included here.

Reply:
Okay, that makes sense if it's controlled in the local regs for the peripheral, and not ccs registers. Seems like a bit of splitting the baby, as maybe at the point that they're telling the clock module what their setting is, they should just do the logic internally and simply need a getter for aclk/pclk

Reply:
@vielster Do we want to keep this as-is, or look at restructuring?

Reply:
I think this probably _should_ be restructured to match the STM32 implementation when it gets merged per #1191 but this platform is separated enough from normal use in this tree that I don't think it's an issue to leave as-is.  It will increase codesize a small amount.